### PR TITLE
Assignment5

### PR DESCRIPTION
**Pull request by kaykay38**
Collaborated with Tyler Jones @worseTyler

### CODE REVIEW

**Comment by OkayNiceBrian** [comment]
I'll review this.

**Review by OkayNiceBrian — COMMENTED** [review]
Update the API project with the following
Add a new Dto namespace ✔
The UserController should only accept or return classes from the Dto namespace. Create any needed classes there and update the controller to use them. You will need to convert between the DTO classes and the relevant entities as needed. ✔
Add OpenAPI (swagger) endpoint ✔
Use the OpenAPI to generate a C# client with relevant interfaces ✔
Decorate all relevant API endpoints in the UserController with the appropriate ProducesResponseType ✔
Update the Web project with the following
Register the generated UserClients class using its interface with the Dependency Injection (DI) service. ✔
Update the UserController to take in the generated client as a dependency. ✔
Convert the UserController from using the MockData to using the client to get data from the API service. ✔
Update the controller to convert back and forth between the ViewModel classes and the API DTO classes as needed. ✔
Update the SecretSanta.Api.Tests with the following:
The controller tests should be interation tests not unit tests. ✔
All endpoints should be tested. ✔
The UserController should use a test double for the IUserRepository to avoid needing to retest business and data logic. ✔
Create a new SecretSanta.Web.Tests projecs with the following:
It should be inside of the test directory. ✔
It should have nullable reference types enabled. ✔
It should have functional tests for all endpoints in the UserController that access the API. ✔
It should use a test double for the API service. This test double should be used to validate the controller's interaction with the API without actually calling the API. ✔

**Comment by kaykay38** [comment]
- Update the API project with the following
  - Add a new Dto namespace ✔
  - The `UserController` should only accept or return classes from the Dto namespace. Create any needed classes there and update the controller to use them. You will need to convert between the DTO classes and the relevant entities as needed. ✔
  - Add OpenAPI (swagger) endpoint ✔
  - Use the OpenAPI to generate a C# client with relevant interfaces ✔
  - Decorate all relevant API endpoints in the `UserController` with the appropriate [ProducesResponseType](https://docs.microsoft.com/en-us/aspnet/core/web-api/action-return-types?view=aspnetcore-5.0#iactionresult-type) ✔
- Update the Web project with the following
  - Register the generated `UserClients` class using its interface with the Dependency Injection (DI) service. ✔
  - Update the `UserController` to take in the generated client as a dependency. ✔
  - Convert the `UserController` from using the `MockData` to using the client to get data from the API service. ✔
  - Update the controller to convert back and forth between the ViewModel classes and the API DTO classes as needed. ✔
- Update the SecretSanta.Api.Tests with the following:
  - The controller tests should be [interation tests](https://docs.microsoft.com/en-us/aspnet/core/test/integration-tests?view=aspnetcore-5.0) not unit tests. ✔
  - All endpoints should be tested. ✔
  - The `UserController` should use a test double for the `IUserRepository` to avoid needing to retest business and data logic. ✔
- Create a new SecretSanta.Web.Tests projecs with the following:
  - It should be inside of the test directory. ✔
  - It should have nullable reference types enabled. ✔
  - It should have functional tests for all endpoints in the `UserController` that access the API. ✔
  - It should use a test double for the API service. This test double should be used to validate the controller's interaction with the API without actually calling the API. ✔

Extra Credit:
- Rather than coping data between view models and DTOs, and DTOs and entities; use [AutoMapper](https://automapper.org/) to handle it for you. ❌

File: SecretSanta/src/SecretSanta.Api/Controllers/UsersController.cs [hunk 1]
```diff
@@ -43,17 +57,38 @@ public ActionResult Delete(int id)
...
+
+            return userDto;
         }
 
         [HttpPut("{id}")]
-        public ActionResult Put(int id, [FromBody] User? user)
+        [ProducesResponseType(StatusCodes.Status400BadRequest)]
+        [ProducesResponseType(StatusCodes.Status404NotFound)]
+        [ProducesResponseType(StatusCodes.Status200OK)]
+        public ActionResult Put(int id, [FromBody] UserDto? user)
```
Comment (Keboo) [inline SecretSanta/src/SecretSanta.Api/Controllers/UsersController.cs:91]: This works but has the same problem as when we had the `SecretSanta.Data.User` as the parameter. This tells the caller that they are to pass in an object with an, Id, FirstName, and LastName properties, however only two of those properties end up getting used.

File: SecretSanta/test/SecretSanta.Web.Tests/Controllers/UsersControllerTests.cs [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+        [TestMethod]
+        public async Task CreatePost_WithValidModelState_InvokesPostAsync()
+        {
+            TestableUsersClient usersClient = Factory.Client;
+            HttpClient client = Factory.CreateClient();
+
+            // UserDto user = new(){ FirstName = "Tyler", LastName = null, Id = null };
+            // string json = System.Text.Json.JsonSerializer.Serialize(user);
+            // StringContent content = new(json);
+            // We did this in class to make a json string but when I used it it didn't seem to be actually sending it to the web controllers Create parameters
```
Comment (Keboo) [inline SecretSanta/test/SecretSanta.Web.Tests/Controllers/UsersControllerTests.cs:46]: Correct, the StringContent approach works for the API controllers, but the web controllers are expecting a form to post data, so we use the FormUrlEncodedContent for those.

File: SecretSanta/test/SecretSanta.Web.Tests/Controllers/UsersControllerTests.cs [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+            Assert.AreEqual<int>(0, usersClient.DeleteAsyncInvocationCount);
+            Assert.IsTrue(usersClient.DeleteAsyncUsersList
+                     .Select(item => item.FirstName)
+                     .Contains("Bad"));
+        }
+
+        // I have no clue how to test such that the ModelState would be invalid
+        // However, there are no instances where ModelState is invalid and our Web and Api projects
+        // communicated with each other so I don't think it's super important considering what we
+        // are testing here
```
Comment (Keboo) [inline SecretSanta/test/SecretSanta.Web.Tests/Controllers/UsersControllerTests.cs:159]: Yep :)

**Review by Keboo — COMMENTED** [review]
- Update the API project with the following
  - Add a new Dto namespace ✔
  - The `UserController` should only accept or return classes from the Dto namespace. Create any needed classes there and update the controller to use them. You will need to convert between the DTO classes and the relevant entities as needed. ✔
  - Add OpenAPI (swagger) endpoint ✔
  - Use the OpenAPI to generate a C# client with relevant interfaces ✔
  - Decorate all relevant API endpoints in the `UserController` with the appropriate [ProducesResponseType](https://docs.microsoft.com/en-us/aspnet/core/web-api/action-return-types?view=aspnetcore-5.0#iactionresult-type) ✔
- Update the Web project with the following
  - Register the generated `UserClients` class using its interface with the Dependency Injection (DI) service. ✔
  - Update the `UserController` to take in the generated client as a dependency. ✔
  - Convert the `UserController` from using the `MockData` to using the client to get data from the API service. ✔
  - Update the controller to convert back and forth between the ViewModel classes and the API DTO classes as needed. ✔
- Update the SecretSanta.Api.Tests with the following:
  - The controller tests should be [integration tests](https://docs.microsoft.com/en-us/aspnet/core/test/integration-tests?view=aspnetcore-5.0) not unit tests. ✔
  - All endpoints should be tested.✔
  - The `UserController` should use a test double for the `IUserRepository` to avoid needing to retest business and data logic.✔
- Create a new SecretSanta.Web.Tests projecs with the following:
  - It should be inside of the test directory. ✔
  - It should have nullable reference types enabled. ✔
  - It should have functional tests for all endpoints in the `UserController` that access the API. ✔
  - It should use a test double for the API service. This test double should be used to validate the controller's interaction with the API without actually calling the API. ✔

**Pull request closed without merging**